Fix the behavior of addrConn with empty address list

Please answer these questions before submitting your issue.

### What version of gRPC are you using?

git head

### What version of Go are you using (`go version`)?

1.9

### What operating system (Linux, Windows, …) and version?

linux

### What did you do?

I've written my own resolver that resolves addresses using serf. When shutting down I eventually call resolver.ClientConn.NewAddress() with an empty array. This is causing a divide by zero in clientconn.go.

### What did you expect to see?

### What did you see instead?

```
I1122 12:28:00.752952    5823 resolver_conn_wrapper.go:106] [GPRC] ccResolverWrapper: sending new addresses to cc: []
I1122 12:28:00.752967    5823 clientconn.go:750] [GPRC] addrConn: tryUpdateAddrs curAddr: {192.168.1.7:33570 0  <nil>}, addrs: []
I1122 12:28:00.752974    5823 clientconn.go:763] [GPRC] addrConn: tryUpdateAddrs curAddrFound: false
I1122 12:28:00.752996    5823 pickfirst.go:68] [GPRC] pickfirstBalancer: HandleSubConnStateChange: 0xc4204ccb60, CONNECTING
panic: runtime error: integer divide by zero

goroutine 3479 [running]:
swarm/vendor/google.golang.org/grpc.(*addrConn).resetTransport(0xc44bb8d400, 0x2540be400, 0x0)
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/clientconn.go:920 +0xfa0
swarm/vendor/google.golang.org/grpc.(*addrConn).connect.func1(0xc44bb8d400)
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/clientconn.go:728 +0x2f
created by swarm/vendor/google.golang.org/grpc.(*addrConn).connect
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/clientconn.go:727 +0xeb
exit status 2
```

I'm getting another crash when I call cc.NewServiceConfig() before any calls to NewAddress()

```
I1122 13:39:00.510892   23302 swarm.go:55] Starting swarm version <unknown> (<unknown>) go1.9
I1122 13:39:00.512396   23302 node.go:162] [NEST] serf agent binding to address 0.0.0.0:37017
I1122 13:39:00.512425   23302 node.go:178] [NEST] serf agent joining cluster at localhost:7946
I1122 13:39:00.514699   23302 resolver_conn_wrapper.go:65] [GPRC] dialing to target with scheme: "serf"
I1122 13:39:00.514920   23302 resolver.go:38] [NEST|resolver] building resolver for {serf  nest}
I1122 13:39:00.515056   23302 resolver_conn_wrapper.go:114] [GPRC] ccResolverWrapper: got new service config: {
		"loadBalancingPolicy": "round_robin"
	}
I1122 13:39:00.515137   23302 clientconn.go:649] [GPRC] ClientConn switching balancer to "round_robin"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xdae039]

goroutine 98 [running]:
swarm/vendor/google.golang.org/grpc.(*ccBalancerWrapper).close(...)
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/balancer_conn_wrappers.go:154
swarm/vendor/google.golang.org/grpc.(*ClientConn).switchBalancer(0xc42015e780, 0xc4202ca0e0, 0xb)
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/clientconn.go:662 +0x139
swarm/vendor/google.golang.org/grpc.(*ClientConn).handleServiceConfig(0xc42015e780, 0x192bacc, 0x2b, 0x1, 0x1)
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/clientconn.go:820 +0x11f
swarm/vendor/google.golang.org/grpc.(*ccResolverWrapper).watcher(0xc4202a67b0)
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/resolver_conn_wrapper.go:115 +0x2d6
created by swarm/vendor/google.golang.org/grpc.newCCResolverWrapper
	/home/predmond/src/swarm/go/src/swarm/vendor/google.golang.org/grpc/resolver_conn_wrapper.go:84 +0x2e1
exit status 2
```

--- a/balancer_conn_wrappers.go
+++ b/balancer_conn_wrappers.go
@@ -19,6 +19,7 @@
 package grpc
 
 import (
+	"fmt"
 	"sync"
 
 	"google.golang.org/grpc/balancer"
@@ -183,6 +184,9 @@ func (ccb *ccBalancerWrapper) handleResolvedAddrs(addrs []resolver.Address, err 
 }
 
 func (ccb *ccBalancerWrapper) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if len(addrs) <= 0 {
+		return nil, fmt.Errorf("grpc: cannot create SubConn with empty address list")
+	}
 	ac, err := ccb.cc.newAddrConn(addrs)
 	if err != nil {
 		return nil, err
@@ -223,6 +227,10 @@ type acBalancerWrapper struct {
 func (acbw *acBalancerWrapper) UpdateAddresses(addrs []resolver.Address) {
 	acbw.mu.Lock()
 	defer acbw.mu.Unlock()
+	if len(addrs) <= 0 {
+		acbw.ac.tearDown(errConnDrain)
+		return
+	}
 	if !acbw.ac.tryUpdateAddrs(addrs) {
 		cc := acbw.ac.cc
 		acbw.ac.mu.Lock()

--- a/clientconn.go
+++ b/clientconn.go
@@ -642,6 +642,8 @@ func (cc *ClientConn) handleResolvedAddrs(addrs []resolver.Address, err error) {
 }
 
 // switchBalancer starts the switching from current balancer to the balancer with name.
+//
+// Caller must hold cc.mu.
 func (cc *ClientConn) switchBalancer(name string) {
 	if cc.conns == nil {
 		return
@@ -659,7 +661,9 @@ func (cc *ClientConn) switchBalancer(name string) {
 
 	// TODO(bar switching) change this to two steps: drain and close.
 	// Keep track of sc in wrapper.
-	cc.balancerWrapper.close()
+	if cc.balancerWrapper != nil {
+		cc.balancerWrapper.close()
+	}
 
 	builder := balancer.Get(name)
 	if builder == nil {
@@ -684,6 +688,8 @@ func (cc *ClientConn) handleSubConnStateChange(sc balancer.SubConn, s connectivi
 }
 
 // newAddrConn creates an addrConn for addrs and adds it to cc.conns.
+//
+// Caller needs to make sure len(addrs) > 0.
 func (cc *ClientConn) newAddrConn(addrs []resolver.Address) (*addrConn, error) {
 	ac := &addrConn{
 		cc:    cc,

--- a/clientconn_test.go
+++ b/clientconn_test.go
@@ -30,6 +30,8 @@ import (
 	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/keepalive"
 	"google.golang.org/grpc/naming"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
 	_ "google.golang.org/grpc/resolver/passthrough"
 	"google.golang.org/grpc/test/leakcheck"
 	"google.golang.org/grpc/testdata"
@@ -327,6 +329,41 @@ func TestNonblockingDialWithEmptyBalancer(t *testing.T) {
 	if err := <-dialDone; err != nil {
 		t.Fatalf("unexpected error dialing connection: %s", err)
 	}
+}
+
+func TestResolverServiceConfigBeforeAddressNotPanic(t *testing.T) {
+	defer leakcheck.Check(t)
+	r, rcleanup := manual.GenerateAndRegisterManualResolver()
+	defer rcleanup()
+
+	cc, err := Dial(r.Scheme()+":///test.server", WithInsecure())
+	if err != nil {
+		t.Fatalf("failed to dial: %v", err)
+	}
+	defer cc.Close()
+
+	// SwitchBalancer before NewAddress. There was no balancer created, this
+	// makes sure we don't call close on nil balancerWrapper.
+	r.NewServiceConfig(`{"loadBalancingPolicy": "round_robin"}`) // This should not panic.
+
+	time.Sleep(time.Second) //Sleep to make sure the service config is handled by ClientConn.
+}
+
+func TestResolverEmptyUpdateNotPanic(t *testing.T) {
+	defer leakcheck.Check(t)
+	r, rcleanup := manual.GenerateAndRegisterManualResolver()
+	defer rcleanup()
+
+	cc, err := Dial(r.Scheme()+":///test.server", WithInsecure())
+	if err != nil {
+		t.Fatalf("failed to dial: %v", err)
+	}
+	defer cc.Close()
+
+	// This make sure we don't create addrConn with empty address list.
+	r.NewAddress([]resolver.Address{}) // This should not panic.
+
+	time.Sleep(time.Second) //Sleep to make sure the service config is handled by ClientConn.
 }
 
 func TestClientUpdatesParamsAfterGoAway(t *testing.T) {

--- a/clientconn_test.go
+++ b/clientconn_test.go
@@ -346,7 +346,7 @@ func TestResolverServiceConfigBeforeAddressNotPanic(t *testing.T) {
 	// makes sure we don't call close on nil balancerWrapper.
 	r.NewServiceConfig(`{"loadBalancingPolicy": "round_robin"}`) // This should not panic.
 
-	time.Sleep(time.Second) //Sleep to make sure the service config is handled by ClientConn.
+	time.Sleep(time.Second) // Sleep to make sure the service config is handled by ClientConn.
 }
 
 func TestResolverEmptyUpdateNotPanic(t *testing.T) {
@@ -363,7 +363,7 @@ func TestResolverEmptyUpdateNotPanic(t *testing.T) {
 	// This make sure we don't create addrConn with empty address list.
 	r.NewAddress([]resolver.Address{}) // This should not panic.
 
-	time.Sleep(time.Second) //Sleep to make sure the service config is handled by ClientConn.
+	time.Sleep(time.Second) // Sleep to make sure the service config is handled by ClientConn.
 }
 
 func TestClientUpdatesParamsAfterGoAway(t *testing.T) {

